canary/mqtt5/canary.ts

```diff
@@ -13,207 +13,263 @@ type Args = { [index: string]: any };
+{
+    return Math.floor(Math.random() * clients.length);
+}
+
+function createCanaryClients(testContext: TestContext, mqttStats: CanaryMqttStatistics): mqtt5.Mqtt5Client[] {
+    const client_config: mqtt5.Mqtt5ClientConfig = {
+        hostName: testContext.hostname,
+        port: testContext.port
     };
 
-    let client : mqtt5.Mqtt5Client = new mqtt5.Mqtt5Client(client_config);
+    const clients = [];
 
-    client.on('error', (error: ICrtError) => {});
-    client.on("messageReceived",(message: mqtt5_packet.PublishPacket) : void => {
-        mqttStats.publishesReceived++;
-    });
+    for (let i = 0; i < testContext.clients; i++) {
+        let client: mqtt5.Mqtt5Client = new mqtt5.Mqtt5Client(client_config);
+
+        client.on('error', (error: ICrtError) => { });
+        client.on("messageReceived", (eventData: mqtt5.MessageReceivedEvent): void => {
+            mqttStats.publishesReceived++;
+        });
+
+        ++mqttStats.clientsUsed;
 
-    return client;
+        clients.push(client);
+    }
+
+    return clients;
 }
 
-async function doSubscribe(context : CanaryContext) {
+async function doSubscribe(context: CanaryContext) {
     let topicFilter: string = `Mqtt5/Canary/RandomSubscribe${uuid()}`;
 
+    let index = getRandomIndex(context.clients);
     try {
         context.mqttStats.subscribesAttempted++;
 
-        await context.client.subscribe({
+        await context.clients[index].subscribe({
             subscriptions: [
-                {topicFilter: RECEIVED_TOPIC, qos: mqtt5_packet.QoS.AtLeastOnce}
+                { topicFilter: RECEIVED_TOPIC, qos: mqtt5.QoS.AtLeastOnce }
             ]
         });
 
-        context.subscriptions.push(topicFilter);
+        context.subscriptions[index].push(topicFilter);
         context.mqttStats.subscribesSucceeded++;
     } catch (err) {
         context.mqttStats.subscribesFailed++;
-        context.subscriptions.filter(entry => entry !== topicFilter);
+        if(context.subscriptions[index].length > 0 )
+            context.subscriptions[index].filter(entry => entry !== topicFilter);
     }
 }
 
-async function doUnsubscribe(context : CanaryContext) {
+async function doUnsubscribe(context: CanaryContext) {
     if (context.subscriptions.length == 0) {
         return;
     }
 
-    let topicFilter: string = context.subscriptions.pop() ?? "canthappen";
+    let index = getRandomIndex(context.clients);
+    let topicFilter: string = context.subscriptions[index].pop() ?? "canthappen";
 
     try {
         context.mqttStats.unsubscribesAttempted++;
 
-        await context.client.unsubscribe({
-            topicFilters: [ topicFilter ]
+        await context.clients[index].unsubscribe({
+            topicFilters: [topicFilter]
         });
 
         context.mqttStats.unsubscribesSucceeded++;
     } catch (err) {
         context.mqttStats.unsubscribesFailed++;
-        context.subscriptions.push(topicFilter);
+        context.subscriptions[index].push(topicFilter);
```

Comment:
I think the check on the line 134 should be `if (context.subscriptions[index].length == 0)`. Otherwise, we're going to add to `context.subscriptions[index]` `"canthappen"` topicfilter.